### PR TITLE
OTP programming Tcl command implementation

### PR DESCRIPTION
**Pull request by chernyee-rs**
> ### Motivate of the pull request
> - [ ] To address an existing issue. If so, please provide a link to the issue: <issue id>
> - [x] Breaking new feature. If so, please describe details in the description part.

This change adds a new OTP programming sub-command `otp` to the programmer command. This `otp` sub-command will be used to program the OTP memory of a device. The command args and options are the same like the existing `fpga_config` sub-command.

Implemented programmer tcl command:-
- programmer otp -c 1 -y \<bitstream filepath\> 

**NOTE**: As OTP programming is not reversable, the `-y` flag must be used to indicate the consensus to proceed with the OTP programming. Omitting the `-y` flag will result in the command stops and return immediately.

### Testing
Without `-y` flag, a warning message is printed on the console:-
image

With `-y` flag, the OTP programming proceed successfully:- 
image

> ### Describe the technical details
> #### What is currently done? (Provide issue link if applicable)
> 
> 
> 
> 
> 
>
> #### What does this pull request change?
> 
> 
> 
> 
> 

> ### Which part of the code base require a change
> 
> - [x] Library: <Specify the library name>
> - [ ] Plug-in: <Specify the plugin name>
> - [ ] Engine
> - [ ] Documentation
> - [ ] Regression tests
> - [ ] Continuous Integration (CI) scripts

> ### Impact of the pull request

> - [ ] Require a change on Quality of Results (QoR)
> - [ ] Break back-compatibility. If so, please list who may be influenced.

### CODE REVIEW

File: src/Configuration/Programmer/Programmer.h [hunk 1]
```diff
@@ -43,6 +43,7 @@ enum ProgrammerErrorCode {
   BitfileNotFound = -108,
   FailedToProgramFPGA = -109,
   OpenOCDExecutableNotFound = -110,
+  FailedToProgramOTP = -111,
   InvalidFlashSize = 111,
```
Comment (guanyung-rs) [inline src/Configuration/Programmer/Programmer.h:47]: Could you help to update these error code to -112 and -113? I think this is something I missed in my previous PR.

Comment (chernyee-rs) [reply src/Configuration/Programmer/Programmer.h]: done.

File: src/Configuration/Programmer/Programmer_helper.cpp [hunk 2]
```diff
@@ -376,6 +376,19 @@ std::string buildFpgaProgramCommand(const Cable& cable, const Device& device,
   return cmd.str() + buildInitEndStringWithCommand(programCommand.str());
 }
 
+std::string buildOTPProgramCommand(const Cable& cable, const Device& device,
```
Comment (guanyung-rs) [inline src/Configuration/Programmer/Programmer_helper.cpp:379]: I think you can add unit test for this function that is similar to `buildFpgaProgramCommand` in tests/unittest/CFGProgrammer/CFGProgrammer_test.cpp file

File: src/Configuration/Programmer/Programmer.cpp [hunk 3]
```diff
@@ -221,6 +227,49 @@ void programmer_entry(const CFGCommon_ARG* cmdarg) {
...
         return;
       }
+    } else if (subCmd == "otp") {
+      auto otp_arg =
+          static_cast<const CFGArg_PROGRAMMER_OTP*>(arg->get_sub_arg());
+      if (otp_arg->confirm == false) {
+        std::string prompt;
+        std::cout << "This OTP programming is not reversable." << std::endl
+                  << "Do you want to continue? [Y/n]";
+        std::cin >> prompt;
```
Comment (esther-rs) [inline src/Configuration/Programmer/Programmer.cpp:237]: n should be capital, consistent with Y

**Review by esther-rs — APPROVED** [review]

File: src/Configuration/Programmer/Programmer.cpp (hunk 3 above)
Comment (ravic-rs) [reply src/Configuration/Programmer/Programmer.cpp:237]: @alain-rs is this the right way to get user input for tcl commands ?

**Review by ravic-rs — CHANGES_REQUESTED** [review]
Blocking for couple of comments to resolve.

File: src/Configuration/Programmer/Programmer.cpp (hunk 3 above)
Comment (chernyee-rs) [reply src/Configuration/Programmer/Programmer.cpp:237]: This user prompting to proceed mechanism is for the interactive sessions (e.g., batch mode & FOEDAG/Raptor UI mode) where user interaction is expected. For the concern of automation (e.g., scripting mode), the command provides a -y flag to gather consensus to proceed and the prompting will be bypassed. (similar UX like Ubuntu apt install -y option).

FYI, this is one of the user requirements from FW team as OTP programming is irreversible and we need to handle this with care for the end-user or designer.

Comment (ravic-rs) [reply src/Configuration/Programmer/Programmer.cpp:237]: Thanks for explaining that, please update the help to reflect these changes and flag.

Comment (KochynVolodymyr) [reply src/Configuration/Programmer/Programmer.cpp:237]: @ravic-rs this is not working in the GUI mode

Comment (chernyee-rs) [reply src/Configuration/Programmer/Programmer.cpp:237]: Programmer GUI will use -y option to bypass the prompting. I will update the Programmer GUI spec to support this feature. Stay tuned.

File: tests/unittest/CFGProgrammer/CFGProgrammer_test.cpp [hunk 4]
```diff
@@ -893,4 +925,88 @@ TEST_F(HwDevicesTestFixture, FindDeviceByNameTest) {
...
+  auto arg = std::make_shared<CFGArg_PROGRAMMER>();
+  arg->parse(argc, argv, nullptr);
+  cmdarg.arg = arg;
+  cmdarg.compilerName = "openfpga";
+  programmer_entry(&cmdarg);
+
+  // just for code coverage but nothing to be observed for now
+}
+
+TEST_F(ProgramerTestFixture, ExecuteOTPCommandWithPromptingAnd_N_Anwser) {
```
Comment (ravic-rs) [inline tests/unittest/CFGProgrammer/CFGProgrammer_test.cpp:984]: Typo here and other tests where this is used.
```suggestion
TEST_F(ProgramerTestFixture, ExecuteOTPCommandWithPromptingAnd_N_Answer) {
```

**Comment by ravic-rs** [comment]
@chernyee-rs this is still missing update to help.txt and description says 
image

Please stay on top of your PR.

File: tests/unittest/CFGProgrammer/CFGProgrammer_test.cpp (hunk 4 above)
Comment (chernyee-rs) [reply tests/unittest/CFGProgrammer/CFGProgrammer_test.cpp]: These UTs are moved.

File: src/Configuration/Programmer/Programmer.cpp (hunk 3 above)
Comment (chernyee-rs) [reply src/Configuration/Programmer/Programmer.cpp]: The user prompting is removed. Users must use -y flag with the OTP command to give consensus to proceed with the OTP programming. Without the -y flag, the command will stop and return immediately. Thus, with this way the command will not create unnecessary blocking if user forget the to use -y flag in scripting mode.

**Review by ravic-rs — APPROVED** [review]

**Comment by ravic-rs** [comment]
@KochynVolodymyr FYI, we will have to update the GUI to include this new "otp" spec, TBD.

**Pull request merged**